pkg/apis/specs/v1alpha2/http_route.go

```diff
@@ -0,0 +1,86 @@
+package v1alpha2
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// HTTPRouteGroup is used to describe HTTP/1 and HTTP/2 traffic.
+// It enumerates the routes that can be served by an application.
+type HTTPRouteGroup struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Most recently observed status of the object.
+	// This data may not be up to date.
+	// Populated by the system.
+	// Read-only.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status Status `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+
+	// Routes for inbound traffic
+	Matches []HTTPMatch `json:"matches,omitempty" protobuf:"bytes,3,opt,name=matches"`
+}
+
+// HTTPMatch defines an individual route for HTTP traffic
+type HTTPMatch struct {
+	// Name is the name of the match for referencing in a TrafficTarget
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+
+	// Methods for inbound traffic as defined in RFC 7231
+	// https://tools.ietf.org/html/rfc7231#section-4
+	Methods []string `json:"methods,omitempty" protobuf:"bytes,1,opt,name=methods"`
+
+	// PathRegex is a regular expression defining the route
+	PathRegex string `json:"pathRegex,omitempty" protobuf:"bytes,1,opt,name=pathRegex"`
+
+	// Headers is a list of headers used to match HTTP traffic
+	Headers httpHeaders `json:"headers,omitempty" protobuf:"bytes,4,opt,name=headers"`
```

Comment:
Why not do `Headers map[string]string` and delete `(h *httpHeaders) UnmarshalJSON`, the codegen knows how to deal with maps deep copy, here is an example: https://github.com/weaveworks/flagger/blob/master/pkg/apis/flagger/v1beta1/canary.go#L298

Reply:
The reason that I created a separate type and `UnmarshalJSON` is because the way that the [smi spec](https://github.com/deislabs/smi-spec/blob/master/traffic-specs.md#http-header-filters) defines how the user should specify headers is as follows:
```
...
headers:
  - user-agent: ".*Android.*"
  - cookie: "^(.*?;)?(type=insider)(;.*)?$"
```

This corresponds to an array of key/value pairs (json looks like: `[{"user-agent":".*Android.*"},{"cookie":"^(.*?;)?(type=insider)(;.*)?$"}]` ). This is something that a `map[string]string` cannot handle by default. So, instead I created an `UnmarshalJSON` with the `httpHeaders` type that takes an `[]map[string]string` and converts it into a single `map[string]string`. I figure that this would be an easier data type to work with than an array of maps.

Reply:
I wrote the spec and I failed to see that haha, ok all good.